pilot/pkg/networking/core/v1alpha3/configgen.go

```diff
@@ -24,16 +24,24 @@ type ConfigGeneratorImpl struct {
 	// List of plugins that modify code generated by this config generator
 	Plugins []plugin.Plugin
 	Cache   model.XdsCache
+
+	CDSCache *CDSCache
 }
 
 func NewConfigGenerator(plugins []plugin.Plugin, cache model.XdsCache) *ConfigGeneratorImpl {
 	return &ConfigGeneratorImpl{
-		Plugins: plugins,
-		Cache:   cache,
+		Plugins:  plugins,
+		Cache:    cache,
+		CDSCache: &CDSCache{},
 	}
 }
 
 // Called when mesh config is changed.
-func (configgen *ConfigGeneratorImpl) MeshConfigChanged(_ *meshconfig.MeshConfig) {
+func (configgen *ConfigGeneratorImpl) MeshConfigChanged(mesh *meshconfig.MeshConfig) {
 	accessLogBuilder.reset()
+
+	// Reset blackhole and passthrough clusters
+	cb := NewClusterBuilder(nil, &model.PushContext{Mesh: mesh})
+	configgen.CDSCache.SetBlackHoleCluster(cb.buildBlackHoleCluster())
```

Comment:
I think this can race with BuildClusters. Not a -race detected race, but we can accidentally revert the changes.

Reply:
There is a lock in CDSCache.

Reply:
Yes its not a data race but you compute the result THEN acquire the lock. Without computing the resource inside the lock you may write out of order. See the "token" stuff in XdsCache it was done to solve similar issue

Reply:
the concern is: compute v1, compute V2, write V2, write v1